packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -0,0 +1,280 @@
+     Aop.apply(X_in,R_in,Teuchos::NO_TRANS, one, zero);
+     R_in.update(one,B_in,negone);
+     return;
+  }
+  const CrsMatrix<SC,LO,GO,NO> & A = *Apt;
+
+  using import_type = typename CrsMatrix<SC,LO,GO,NO>::import_type;
+  using export_type = typename CrsMatrix<SC,LO,GO,NO>::export_type;
+  using MV = MultiVector<SC,LO,GO,NO>;
+
+  // We treat the case of a replicated MV output specially.
+  const bool R_is_replicated =
+    (! R_in.isDistributed () && A.getComm ()->getSize () != 1);
+
+  // It's possible that R is a view of X or B.  
+  // We don't try to to detect the more subtle cases (e.g., one is a
+  // subview of the other, but their initial pointers differ).  We
+  // only need to do this if this matrix's Import is trivial;
+  // otherwise, we don't actually apply the operator from X into Y.
+  
+  RCP<const import_type> importer = A.getGraph ()->getImporter ();
+  RCP<const export_type> exporter = A.getGraph ()->getExporter ();
+
+  // Temporary MV for Import operation.  After the block of code
+  // below, this will be an (Imported if necessary) column Map MV
+  // ready to give to localApply(...).
+  RCP<const MV> X_colMap;
+  if (importer.is_null ()) {
+    if (! X_in.isConstantStride ()) {
+      // Not all sparse mat-vec kernels can handle an input MV with
+      // nonconstant stride correctly, so we have to copy it in that
+      // case into a constant stride MV.  To make a constant stride
+      // copy of X_in, we force creation of the column (== domain)
+      // Map MV (if it hasn't already been created, else fetch the
+      // cached copy).  This avoids creating a new MV each time.
+      RCP<MV> X_colMapNonConst = A.getColumnMapMultiVector (X_in, true);
+      Tpetra::deep_copy (*X_colMapNonConst, X_in);
+      X_colMap = rcp_const_cast<const MV> (X_colMapNonConst);
+    }
+    else {
+      // The domain and column Maps are the same, so do the local
+      // multiply using the domain Map input MV X_in.
+      X_colMap = rcpFromRef (X_in);
+    }
+  }
+  else { // need to Import source (multi)vector
+    ProfilingRegion regionImport ("Tpetra::CrsMatrix::residual: Import");
+    // We're doing an Import anyway, which will copy the relevant
+    // elements of the domain Map MV X_in into a separate column Map
+    // MV.  Thus, we don't have to worry whether X_in is constant
+    // stride.
+    RCP<MV> X_colMapNonConst = A.getColumnMapMultiVector (X_in);
+    
+    // Import from the domain Map MV to the column Map MV.
+    X_colMapNonConst->doImport (X_in, *importer, INSERT);
+    X_colMap = rcp_const_cast<const MV> (X_colMapNonConst);
+  }
+
+  // Get a vector for the rowMap output residual
+  RCP<MV> R_rowMap;
+  if(exporter.is_null()) {
+    R_rowMap = rcpFromRef(R_in);
+  }
+  else {
+    R_rowMap = A.getRowMapMultiVector (R_in);    
+  }
+  
+  // FIXME: Need to get a potential cached buffer for B (this will need to be another function)
```

Comment:
I expect the common case to be row Map == range Map, so it's OK to create a temp vector otherwise.  Also need to create if not constant stride.

Reply:
Got it